internal/transport/grpchttp2/framer.go

```diff
@@ -0,0 +1,252 @@
+	SettingsInitialWindowSize    SettingID = 0x4
+	SettingsMaxFrameSize         SettingID = 0x5
+	SettingsMaxHeaderListSize    SettingID = 0x6
+)
+
+// FrameHeader is the 9 byte header of any HTTP/2 Frame.
+// See [Frame Header].
+//
+// [Frame Header]: https://httpwg.org/specs/rfc7540.html#FrameHeader
+type FrameHeader struct {
+	// Size is the size of the frame's payload without the 9 header bytes.
+	// As per the HTTP/2 spec, size can be up to 3 bytes, but only frames
+	// up to 16KB can be processed without agreement.
+	Size uint32
+	// Type is a byte that represents the Frame Type.
+	Type FrameType
+	// Flags is a byte representing the flags set on this Frame.
+	Flags Flag
+	// StreamID is the ID for the stream which this frame is for. If the
+	// frame is connection specific instead of stream specific, the
+	// streamID is 0.
+	StreamID uint32
+}
+
+// Frame represents an HTTP/2 Frame.
+type Frame interface {
+	Header() *FrameHeader
+	// Free frees the underlying buffer if present so it can be reused by the
+	// framer.
+	Free()
+}
+
+type DataFrame struct {
+	hdr  *FrameHeader
+	free func()
+	Data []byte
+}
+
+func (f *DataFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *DataFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type HeadersFrame struct {
+	hdr      *FrameHeader
+	free     func()
+	HdrBlock []byte
+}
+
+func (f *HeadersFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *HeadersFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type RSTStreamFrame struct {
+	hdr  *FrameHeader
+	Code ErrCode
+}
+
+func (f *RSTStreamFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *RSTStreamFrame) Free() {}
+
+type SettingsFrame struct {
+	hdr      *FrameHeader
+	Settings []Setting
+}
+
+func (f *SettingsFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *SettingsFrame) Free() {}
+
+type PingFrame struct {
+	hdr  *FrameHeader
+	free func()
+	Data []byte
+}
+
+func (f *PingFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *PingFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type GoAwayFrame struct {
+	hdr          *FrameHeader
+	free         func()
+	LastStreamID uint32
+	Code         ErrCode
+	DebugData    []byte
+}
+
+func (f *GoAwayFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *GoAwayFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type WindowUpdateFrame struct {
+	hdr *FrameHeader
+	Inc uint32
+}
+
+func (f *WindowUpdateFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+type ContinuationFrame struct {
+	hdr      *FrameHeader
+	free     func()
+	HdrBlock []byte
+}
+
+func (f *ContinuationFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *ContinuationFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+// MetaHeadersFrame is not a Frame Type that appears on the HTTP/2 Spec. It is
+// a representation of the merging and decoding of all the Headers and
+// Continuation frames on a Stream.
```

Comment:
I would rephrase this to say something like

```suggestion
// MetaHeadersFrame is the representation of one HEADERS frame and zero or more
// contiguous CONTINUATION frames and the decoding of their HPACK-encoded
// contents.  This frame type is not transmitted over the network and is only
// generated by the ReadFrame() function.
```